src/library/treeitemmodel.cpp

```diff
@@ -101,64 +103,61 @@ QVariant TreeItemModel::headerData(int section, Qt::Orientation orientation, int
 }
 
 QModelIndex TreeItemModel::index(int row, int column, const QModelIndex &parent) const {
-    if (!hasIndex(row, column, parent))
+    if (!hasIndex(row, column, parent)) {
         return QModelIndex();
+    }
 
-    TreeItem *parentItem = NULL;
-
-    if (!parent.isValid())
-        parentItem = m_pRootItem;
-    else
+    TreeItem *parentItem;
+    if (parent.isValid()) {
         parentItem = static_cast<TreeItem*>(parent.internalPointer());
+    } else {
+        parentItem = getRootItem();
+    }
 
     TreeItem *childItem = parentItem->child(row);
-    if (childItem)
+    if (childItem) {
         return createIndex(row, column, childItem);
-    else
+    } else {
         return QModelIndex();
+    }
 }
 
-QModelIndex TreeItemModel::parent(const QModelIndex &index) const {
-    if (!index.isValid())
+QModelIndex TreeItemModel::parent(const QModelIndex& index) const {
+    if (!index.isValid()) {
         return QModelIndex();
+    }
 
     TreeItem *childItem = static_cast<TreeItem*>(index.internalPointer());
     TreeItem *parentItem = childItem->parent();
-
-    if (parentItem == m_pRootItem)
+    if (parentItem == getRootItem()) {
         return QModelIndex();
-
-    return createIndex(parentItem->row(), 0, parentItem);
+    } else {
+        return createIndex(parentItem->parentRow(), 0, parentItem);
+    }
 }
 
-int TreeItemModel::rowCount(const QModelIndex &parent) const {
-    if (parent.column() > 0)
+int TreeItemModel::rowCount(const QModelIndex& parent) const {
+    if (parent.column() > 0) {
         return 0;
-
-    TreeItem *parentItem = NULL;
-    //qDebug() << "parent data: " << parent.data();
-    if (!parent.isValid()) {
-        parentItem = m_pRootItem;
     }
-    else{
-        parentItem = static_cast<TreeItem*>(parent.internalPointer());
 
+    TreeItem* parentItem;
+    if (parent.isValid()) {
+        parentItem = static_cast<TreeItem*>(parent.internalPointer());
+    } else {
+        parentItem = getRootItem();
     }
-
-    //qDebug() << "TreeItem data: " << parent.internalPointer();
-
-    return parentItem->childCount();
+    return parentItem->childRows();
 }
 
 /**
  * Populates the model and notifies the view.
  * Call this method first, before you do call any other methods.
  */
-void TreeItemModel::setRootItem(TreeItem *item) {
-    if(m_pRootItem) delete m_pRootItem;
-
-    m_pRootItem = item;
+TreeItem* TreeItemModel::setRootItem(std::unique_ptr<TreeItem> pRootItem) {
```

Comment:
I am just thinking about the best way to pass the unique_ptr here. 

I did expect that it has to be "std::unique_ptr<TreeItem>&& pRootItem" just as like a move constructor. 
After reading a bit the current implementation seams also common, but why? 

setRootItem(std::move(root)); 
std::move casts root to an rvalue reference. In the pass by value case, the rvalue reference is moved to the function parameter, emptying "root". Inside setRootItem() the same happens again, cast to rvalue, copy rvalue to value = move.

If we switch to a pass by rvalue reference. We only have a single move action inside setRootItem. 
From the interface the caller has no idea if the move actually happens or not. 
As have read, this is considered as disadvantage.
I do not understand why this is important. This is the same when using move constructors. 
I would just consider a value that was used by std::move() as invalid.

How do you think about it?